account/recovery.go

```diff
@@ -123,17 +128,254 @@ func recoverInitalState(ctx context.Context, cfg RecoveryConfig) (
+	//         - Look for the script in all of the node txs.
+
+	target := cfg.AccountTarget
+	accounts := make([]*Account, 0, target)
+
+	batchKey := cfg.InitialBatchKey
+	// DecrementKey so the first for iteration starts with the
+	// first batchKey value.
+	batchKey = poolscript.DecrementKey(batchKey)
+
+searchLoop:
+	for batchCounter := 0; batchCounter < 5000; batchCounter++ {
+		batchKey = poolscript.IncrementKey(batchKey)
+		for h := cfg.FirstBlock; h < cfg.LastBlock; h++ {
+		nextAccount:
+			for acc := range possibleAccounts {
+				script, err := poolscript.AccountScript(
+					h,
+					acc.TraderKey.PubKey,
+					cfg.AuctioneerPubKey,
+					batchKey,
+					acc.Secret,
+				)
+				if err != nil {
+					key := acc.TraderKey.PubKey
+					log.Debugf(
+						"unable to generate script: \n"+
+							"height: %v\n"+
+							"batchKey: %v\n"+
+							"traderKey: %v\n",
+						h,
+						batchKey,
+						hex.EncodeToString(
+							key.SerializeCompressed(),
+						),
+					)
+					goto nextAccount
+				}
+
+				if tx, idx := appearsInTxs(
+					acc, script, cfg.Transactions,
+				); tx != nil {
+					// If it's a match, populate
+					// the account information.
+					acc.Expiry = h
+					acc.Value = btcutil.Amount(
+						tx.Tx.TxOut[idx].Value,
+					)
+					acc.BatchKey = batchKey
+					acc.OutPoint = wire.OutPoint{
+						Hash:  tx.Tx.TxHash(),
+						Index: idx,
+					}
+					acc.LatestTx = tx.Tx
+					acc.State = StateOpen
+
+					accounts = append(accounts, acc)
+
+					delete(possibleAccounts, acc)
+
+					// If we already found all the
+					// accounts that we were looking
+					// for quit the search.
+					if len(accounts) == int(target) {
+						break searchLoop
+					}
+				}
+			}
+		}
+	}
+
+	return accounts
+}
+
+// appearsInTxs tries to locate a script in any of the provided transactions.
+func appearsInTxs(acc *Account, script []byte, txs []lndclient.Transaction) (
+	*lndclient.Transaction, uint32) {
+
+	for _, tx := range txs {
+		idx, ok := poolscript.LocateOutputScript(tx.Tx, script)
+		if ok {
+			tx := tx // make scopelint happy
+			traderKey := acc.TraderKey.PubKey.SerializeCompressed()
+			log.Debugf(
+				"found accout with trader key %v",
+				hex.EncodeToString(traderKey),
+			)
+			return &tx, idx
+		}
+	}
+	return nil, 0
+}
+
 // updateAccountStates tries to update the states for every provided
 // account up to their latest state by following the on chain
 // modification footprints.
-func updateAccountStates(cfg RecoveryConfig, accounts []*Account) (
-	[]*Account, error) {
+func updateAccountStates(cfg RecoveryConfig, accounts []*Account) []*Account {
 
 	recoveredAccounts := make([]*Account, 0, len(accounts))
 
-	// TODO (positiveblue): update account states
+nextAccount:
+	for _, acc := range accounts {
+		log.Debugf(
+			"Updating state for account with trader key %v",
+			hex.EncodeToString(
+				acc.TraderKey.PubKey.SerializeCompressed(),
+			),
+		)
+		for _, tx := range cfg.Transactions {
```

Comment:
This will work as long as `lnd` wasn't also recovered from seed. If the node was recovered from the seed then the initial account funding transaction will be found on chain because it was funded with lnd wallet coins. But all subsequent account updates won't be recognized since lnd doesn't know the possible scripts.
That's why we'd need to follow the spends on chain. But I'm okay with us defining that as an edge case for now.